gax/src/main/java/com/google/api/gax/retrying/ExponentialRetryAlgorithm.java

```diff
@@ -135,9 +135,21 @@ public boolean shouldRetry(TimedAttemptSettings nextAttemptSettings) {
             - nextAttemptSettings.getFirstAttemptStartTimeNanos()
             + nextAttemptSettings.getRandomizedRetryDelay().toNanos();
 
-    return totalTimeSpentNanos <= globalSettings.getTotalTimeout().toNanos()
-        && (globalSettings.getMaxAttempts() <= 0
-            || nextAttemptSettings.getAttemptCount() < globalSettings.getMaxAttempts());
+    int maxAttempts = globalSettings.getMaxAttempts();
+    long totalTimeout = globalSettings.getTotalTimeout().toNanos();
+
+    // If totalTimeout limit is defined, check that it hasn't been crossed
+    if (totalTimeout > 0 && totalTimeSpentNanos > totalTimeout) {
+      return false;
+    }
+
+    // If maxAttempts limit is defined, check that it hasn't been crossed
+    if (maxAttempts > 0 && nextAttemptSettings.getAttemptCount() >= maxAttempts) {
+      return false;
+    }
+
+    // No limits crossed
+    return true;
```

Comment:
Now `no-retries` case is supposed to be configured now? I guess, even if it is still possible to configure no-retries here, the way it is done has changed (it used to be default case). This may cause retries for things, which should not be retried. Which is especially dangerous for non-idempotent calls.

Reply:
The no retry case should stay the same:
https://github.com/googleapis/gax-java/blob/68ae22d90e8d01c1f8996945c64cb502ee699742/gax/src/main/java/com/google/api/gax/rpc/UnaryCallSettings.java#L140-L154

Just set MaxAttempts = 1

Reply:
Correct, but what if somebody already has it in their code without `maxAttempts = 1`? I guess that config meant "no retries", now it will mean "retry forever".

Reply:
Yes, you are right that would change the behavior. Maybe we can compromise on changing `return true;` to:

```java
// Prevent forever retries
return maxAttempts > 0 || totalTimeout > 0;
```
Does that work?

Reply:
Will disabling retries if both `maxAttempts` & `totalTimeout` are 0 be compatible with the old behavior (I guess it is), then Ok.

Reply:
I believe so